File: frontend/src/stores/presence.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { ApiError, apiGetAuthed } from "../api/client";
import type { Presence } from "../generated/Presence";
import type { PresenceList } from "../generated/PresenceList";
import type { PresenceStatus } from "../generated/PresenceStatus";
import { useAuthStore } from "./auth";
import { useRealtimeStore } from "./realtime";

/**
 * Presence (CONTEXT.md: 在线状态) — who is reachable, and when they last were.
 *
 * One map keyed by **User** id, fed from two directions:
 *
 * - the socket, through `onPresenceEvent`, for a change as it happens. The server
 *   only sends that to Participants of a Conversation shared with the User, so a
 *   client is never told about a stranger;
 * - `GET /presence`, for the current state of the peers a view is about to render.
 *   A live event only exists *when something changes*, so the read is what a
 *   freshly opened screen uses.
 *
 * The state is scoped to the User, not to a Device: a User is online while any of
 * their Devices is, and the server owns that decision. This store never derives
 * presence from a socket of its own.
 *
 * An entry is **replaced**, never merged: an `online` event carries no last-seen
 * instant and an `offline` one carries the instant the User stopped being
 * reachable, so the newest event is the whole truth for that User.
 */
export const usePresenceStore = defineStore("presence", () => {
  /** The known presence of each User, by User id. Absent means "not known". */
  const presences = ref<Record<string, Presence>>({});
  const loading = ref(false);
  const errorMessage = ref<string | null>(null);

  /** Adopt one presence, from the socket or from a REST read. Idempotent. */
  function apply(presence: Presence): void {
    presences.value[presence.user_id] = presence;
  }

  /** The last known status of a User, or `null` when nothing is known yet. */
  function statusFor(userId: string): PresenceStatus | null {
    return presences.value[userId]?.status ?? null;
  }

  /** When the User was last reachable, or `null` while online or unknown. */
  function lastSeenFor(userId: string): number | null {
    return presences.value[userId]?.last_seen_ms ?? null;
  }

  /**
   * Load the current presence of the named Users.
   *
   * Blank ids and duplicates are dropped, and a request with nothing to ask about
   * is not sent at all — the server refuses an empty list, and "nobody to ask
   * about" is not a failure. A User the caller shares no Conversation with is
   * absent from the answer rather than reported offline; this store simply keeps
   * whatever it already knew about them.
   */
  async function load(userIds: readonly string[]): Promise<void> {
    const token = useAuthStore().accessToken;
    if (token === null) return;

    const ids = [...new Set(userIds.filter((userId) => userId !== ""))];
    if (ids.length === 0) return;

    loading.value = true;
    errorMessage.value = null;
    try {
      const query = encodeURIComponent(ids.join(","));
      const payload = await apiGetAuthed<PresenceList>(
        `/presence?user_ids=${query}`,
        token,
      );
      for (const presence of payload.presences) apply(presence);
    } catch (cause) {
      errorMessage.value =
        cause instanceof ApiError
          ? (cause.body?.error.message ?? `请求失败（HTTP ${cause.status}）`)
          : "无法加载在线状态";
    } finally {
      loading.value = false;
    }
  }

  /** Forget every entry; used when the session ends so nothing leaks across accounts. */
  function clear(): void {
    presences.value = {};
  }

  // One subscription for the store's lifetime: the realtime store forwards only
  // presence changes here, and this store is what knows what they mean.
  useRealtimeStore().onPresenceEvent(apply);

  return {
    presences,
    loading,
    errorMessage,
    apply,
    statusFor,
    lastSeenFor,
    load,
    clear,
  };
});

File: backend/crates/contract/src/presence.rs
//! Presence (CONTEXT.md: 在线状态) — who is reachable, and when they last were.
//!
//! Presence has two halves, and this module carries both of them on the wire:
//!
//! - **Live**: a [`Presence`] rides [`crate::ServerEvent::Presence`] to the
//!   Participants of a shared Conversation the moment a User's reachability
//!   changes. The change is always a **User** transition — going online is the
//!   *first* of a User's Devices connecting and going offline is the *last* one
//!   disconnecting (CONTEXT.md draws exactly this Device-vs-User line) — so a
//!   phone and a laptop never flicker each other offline.
//! - **On demand**: [`PresenceList`] is the REST read model, so a client that has
//!   just opened a Conversation can render the current state without waiting for
//!   the next change.
//!
//! # Visibility is a decision, not a fan-out
//!
//! Both halves answer "who may see this User's presence" through the same rule:
//! **only Participants of a shared Conversation**. That rule lives in one seam
//! (`jiuyue_chat::ChatService::presence_audience`), so the "who can see my
//! presence" settings of issue #27 become a *filter applied at that seam* rather
//! than a rewrite of every call site. Nothing here fans out unconditionally.
//!
//! # Last seen
//!
//! [`Presence::last_seen_ms`] is `null` while a User is online (the present is not
//! a past instant) and carries the instant they stopped being reachable when they
//! are offline. It is the one durable half of presence: reachability itself is
//! in-process state, so a restart makes everybody offline by construction, while
//! the last-seen instant survives in PostgreSQL.

use serde::{Deserialize, Serialize};
use ts_rs::TS;

/// A User's reachability (CONTEXT.md: Presence).
///
/// Two states, not three: `away` in CONTEXT.md is an *idle* refinement of online
/// (reachable, just quiet) and is deliberately not produced yet — adding the
/// variant later is an additive contract change, which is the whole reason the
/// variants are a string-tagged enum rather than a boolean.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(export)]
pub enum PresenceStatus {
    /// At least one of the User's Devices is connected.
    Online,
    /// No Device of the User is connected.
    Offline,
}

/// One User's presence, as the server reports it.
///
/// Sent as [`crate::ServerEvent::Presence`] on a change, and listed by
/// `GET /presence` for a client that needs the current state now.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct Presence {
    /// ULID of the User whose presence this is.
    pub user_id: String,
    /// Whether the User is reachable right now.
    pub status: PresenceStatus,
    /// When the User was last reachable, milliseconds since the Unix epoch.
    ///
    /// `null` in two cases, both meaning "no past instant to show": while the
    /// User is online, and for a User who has never been seen offline yet.
    #[ts(type = "number | null")]
    pub last_seen_ms: Option<i64>,
}

/// Query parameters of `GET /presence`.
///
/// `user_ids` is a comma-separated list of ULIDs — the shape a query string can
/// carry without repeated keys. The handler splits, trims and de-duplicates it and
/// refuses more than [`MAX_PRESENCE_QUERY`] entries, so one request can never ask
/// the server to materialise an unbounded answer on the 2 vCPU / 2 GB box.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct PresenceQuery {
    /// Comma-separated ULIDs to ask about; required and non-empty.
    pub user_ids: Option<String>,
}

/// `GET /presence` response: one entry per **visible** requested User.
///
/// A requested User who does not share a Conversation with the caller is omitted
/// rather than reported offline — the list is the visibility-filtered answer, not
/// a directory. Order follows the request.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct PresenceList {
    /// The visible Users' presence, in request order.
    pub presences: Vec<Presence>,
}

/// Upper bound on how many Users one `GET /presence` may ask about.
///
/// Sized like the other list bounds in this contract: generous for a screenful of
/// participants (a Direct peer, a group's member list) while keeping the point
/// lookup's `= ANY($1)` array, and the response, bounded.
pub const MAX_PRESENCE_QUERY: usize = 200;

#[cfg(test)]
mod tests {
    use super::{MAX_PRESENCE_QUERY, Presence, PresenceList, PresenceQuery, PresenceStatus};

    #[test]
    fn a_presence_serialises_with_snake_case_status_and_null_last_seen() {
        let presence = Presence {
            user_id: "01JABC1234567890ABCDEFGHJ3".to_owned(),
            status: PresenceStatus::Online,
            last_seen_ms: None,
        };

        let wire = serde_json::to_value(&presence).expect("presence must serialise");

        assert_eq!(wire["user_id"], "01JABC1234567890ABCDEFGHJ3");
        assert_eq!(wire["status"], "online");
        assert!(wire["last_seen_ms"].is_null());
    }

    #[test]
    fn an_offline_presence_carries_the_instant_the_user_was_last_seen() {
        let presence = Presence {
            user_id: "01JABC1234567890ABCDEFGHJ3".to_owned(),
            status: PresenceStatus::Offline,
            last_seen_ms: Some(1_750_000_000_000),
        };

        let wire = serde_json::to_value(&presence).expect("presence must serialise");

        assert_eq!(wire["status"], "offline");
        assert_eq!(wire["last_seen_ms"], 1_750_000_000_000_i64);
    }

    #[test]
    fn the_query_carries_a_comma_separated_list_and_the_list_wraps_presences() {
        let query: PresenceQuery = serde_json::from_value(serde_json::json!({ "user_ids": "a,b" }))
            .expect("the query must deserialise");
        assert_eq!(query.user_ids.as_deref(), Some("a,b"));

        let list = PresenceList {
            presences: vec![Presence {
                user_id: "01JABC1234567890ABCDEFGHJ3".to_owned(),
                status: PresenceStatus::Offline,
                last_seen_ms: None,
            }],
        };
        let wire = serde_json::to_value(&list).expect("the list must serialise");
        assert_eq!(wire["presences"].as_array().map(Vec::len), Some(1));
    }

    #[test]
    fn the_query_bound_is_a_positive_screenful() {
        // A group's member list is the widest realistic single query; the bound
        // must comfortably hold one and stay finite.
        const {
            assert!(MAX_PRESENCE_QUERY >= 100, "a group member list must fit");
            assert!(
                MAX_PRESENCE_QUERY <= 1_000,
                "the point lookup must stay bounded"
            );
        }
    }
}
